src/scikit_build_core/builder/_load_provider.py

```diff
@@ -0,0 +1,150 @@
+from __future__ import annotations
+
+import dataclasses
+import importlib
+import inspect
+import sys
+from collections.abc import Iterator, Mapping
+from pathlib import Path
+from typing import TYPE_CHECKING, Any, Protocol, Union, runtime_checkable
+
+if TYPE_CHECKING:
+    from collections.abc import Generator, Iterable
+
+    StrMapping = Mapping[str, Any]
+else:
+    StrMapping = Mapping
+
+from ..metadata import _ALL_FIELDS
+
+__all__ = ["load_dynamic_metadata", "load_provider"]
+
+
+def __dir__() -> list[str]:
+    return __all__
+
+
+@runtime_checkable
+class DynamicMetadataProtocol(Protocol):
+    def dynamic_metadata(
+        self,
+        fields: Iterable[str],
+        settings: dict[str, Any],
+        metadata: Mapping[str, Any],
+    ) -> dict[str, Any]: ...
+
+
+@runtime_checkable
+class DynamicMetadataRequirementsProtocol(DynamicMetadataProtocol, Protocol):
+    def get_requires_for_dynamic_metadata(
+        self, settings: dict[str, Any]
+    ) -> list[str]: ...
+
+
+@runtime_checkable
+class DynamicMetadataWheelProtocol(DynamicMetadataProtocol, Protocol):
+    def dynamic_wheel(
+        self, field: str, settings: Mapping[str, Any] | None = None
+    ) -> bool: ...
+
+
+DMProtocols = Union[
+    DynamicMetadataProtocol,
+    DynamicMetadataRequirementsProtocol,
+    DynamicMetadataWheelProtocol,
+]
+
+
+def load_provider(
+    provider: str,
+    provider_path: str | None = None,
+) -> DMProtocols:
+    if provider_path is None:
+        return importlib.import_module(provider)
+
+    if not Path(provider_path).is_dir():
+        msg = "provider-path must be an existing directory"
+        raise AssertionError(msg)
+
+    try:
+        sys.path.insert(0, provider_path)
+        return importlib.import_module(provider)
+    finally:
+        sys.path.pop(0)
+
+
+def _load_dynamic_metadata(
+    metadata: Mapping[str, Mapping[str, str]],
+) -> Generator[tuple[str, DMProtocols, dict[str, str]], None, None]:
+    for field, orig_config in metadata.items():
+        if "provider" not in orig_config:
+            msg = "Missing provider in dynamic metadata"
+            raise KeyError(msg)
+
+        if field not in _ALL_FIELDS:
+            msg = f"{field} is not a valid field"
+            raise KeyError(msg)
+        config = dict(orig_config)
+        provider = config.pop("provider")
+        provider_path = config.pop("provider-path", None)
+        loaded_provider = load_provider(provider, provider_path)
+        yield field, loaded_provider, config
+
+
+@dataclasses.dataclass
+class DynamicSettings(StrMapping):
+    settings: dict[str, dict[str, Any]]
+    project: dict[str, Any]
+    providers: dict[str, DMProtocols]
+
+    def __getitem__(self, key: str) -> Any:
```

Comment:
We should also override `__contains__` so that this does not get called when simple `in` checks are used.

Reply:
Ah, good point. Vaguely thought about that at one point but not when I was implementing it.

Reply:
Yeah, I just got hit by that in the other PR 😅